pennylane/math/jax_box.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the JaxBox implementation of the TensorBox API.
+"""
+import jax.numpy as jnp
+import pennylane as qml
+
+
+wrap_output = qml.math.wrap_output
+
+
+class JaxBox(qml.math.TensorBox):
+    """Implements the :class:`~.TensorBox` API for ``numpy.ndarray``.
+
+    For more details, please refer to the :class:`~.TensorBox` documentation.
+    """
+
+    abs = wrap_output(lambda self: jnp.abs(self.data))
+    angle = wrap_output(lambda self: jnp.angle(self.data))
+    arcsin = wrap_output(lambda self: jnp.arcsin(self.data))
+    cast = wrap_output(lambda self, dtype: jnp.array(self.data, dtype=dtype))
+    expand_dims = wrap_output(lambda self, axis: jnp.expand_dims(self.data, axis=axis))
+    ones_like = wrap_output(lambda self: jnp.ones_like(self.data))
+    sqrt = wrap_output(lambda self: jnp.sqrt(self.data))
+    sum = wrap_output(
+        lambda self, axis=None, keepdims=False: jnp.sum(self.data, axis=axis, keepdims=keepdims)
+    )
+    T = wrap_output(lambda self: self.data.T)
+    take = wrap_output(
+        lambda self, indices, axis=None: jnp.take(self.data, indices, axis=axis, mode="wrap")
+    )
+
+    def __init__(self, tensor):
+        tensor = jnp.asarray(tensor)
+
+        super().__init__(tensor)

```

Comment:
Tangent: this method looks very trivial but it avoids the `__new__` method from being called recursively, since it is inherited!

Reply:
A case of being too clever for our own good.

The main thinking behind the `TensorBox` class was making the internal API both (a) extensible and (b) documented, as opposed to having functions with branches for each framework. So that anyone could easily define the mapping between `qml.math` API and their framework, and it would just 'plug in' to PL, even though the end-user interface is simply functional.

But, I think `__new__` is not the right way of doing it, because someones `Box` implementation requires modifying the PL source code to be plugged in. Maybe a better approach is to provide an entry point, like we do with plugins.